### PR TITLE
Introduce startup diagnostics

### PR DESCRIPTION
**Pull request by andreasohlund**
@Particular/nservicebus-maintainers this is just a first draft, thoughts on the API?

### CODE REVIEW

**Comment by bording** [comment]
@andreasohlund What use cases are you thinking of for `CustomDiagnosticsWriter`?

**Comment by mikeminutillo** [comment]
I would like to see a custom diagnostics writer for ServiceControl :)

I imagine an API along the lines of: 

```
var diagnostics = endpointConfiguration.Diagnostics();

// To not write any to diagnostics
diagnostics.Disable();

// To not write to the log directory
diagnostics.DoNotWriteToLogFolder();

// Write to a custom writer
diagnostics.WriteTo(customWriter);

// To write to a custom file location. This can just be an extension method that calls `WriteTo()`
diagnostics.WriteToFile(location);
```

The only part I'm not happy with is `DoNotWriteToLogFolder()`.

**Comment by timbussmann** [comment]
> I would like to see a custom diagnostics writer for ServiceControl :)

tbh, I'll call yagni on that. Also, that would have to be an additonal diagnostics writer and should not replace the file writer.

File: src/NServiceBus.Core/Diagnostics/EndpointDiagnostics.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            public WriteStartupDiagnostics(DiagnosticsWriter diagnosticsWriter)
+            {
+                this.diagnosticsWriter = diagnosticsWriter;
+            }
+
+            protected override Task OnStart(IMessageSession session)
+            {
+                var data = "tbd";
+
+                return diagnosticsWriter.Write(data);
```
Comment (timbussmann) [inline src/NServiceBus.Core/Diagnostics/EndpointDiagnostics.cs]: definitely needs some error handling

**Comment by andreasohlund** [comment]
>What use cases are you thinking of for CustomDiagnosticsWriter?

I had azure in mind for this since writing to file there is pretty pointless, so you would likely write it to `Trace` or whatever make sense there.

See the discussion here https://github.com/Particular/PlatformDevelopment/issues/1397#issuecomment-331219498 for more details

Does this make sense?

File: src/NServiceBus.Core/Diagnostics/EndpointDiagnostics.cs (hunk 1 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Diagnostics/EndpointDiagnostics.cs]: Just went with a try catch => log to have the endpoint start even if there are issues

**Comment by andreasohlund** [comment]
@SimonCropp @timbussmann I've moved some stuff around and made the addition of diagnostics extendable.

What do you think?

**Comment by andreasohlund** [comment]
Eg. this is how routing info is added https://github.com/Particular/NServiceBus/pull/5004/files#diff-9c5e3e54a5fff62b085abb4b3b7c245dR62

File: src/NServiceBus.AcceptanceTests/ConfigureEndpointLearningTransport.cs [hunk 2]
```diff
@@ -36,6 +36,8 @@ public Task Configure(string endpointName, EndpointConfiguration configuration,
 
         storageDir = Path.Combine(tempDir, testRunId);
 
+        configuration.SetDiagnosticsRootPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, endpointName, testRunId));
```
Comment (SimonCropp) [inline src/NServiceBus.AcceptanceTests/ConfigureEndpointLearningTransport.cs]: Wouldnt we disable for acc tests?

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                this.settings = settings;
+            }
+
+            protected override async Task OnStart(IMessageSession session)
+            {
+                try
+                {
+                    var data = "";
+
+                    foreach (var section in settings.Get<StartupDiagnosticEntries>().Entries)
```
Comment (SimonCropp) [inline src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs]: Is it not possible to serialize the whole dictionary?

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                    {
+                        var sectionData = SimpleJson.SerializeObject(section.Data);
+
+                        if (!string.IsNullOrEmpty(data))
+                        {
+                            data += "," + Environment.NewLine;
```
Comment (SimonCropp) [inline src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs]: Stringbuilder FTW

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs (hunk 3 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs]: I tried and it gave this:

```
{
   [
      {
         "Name":"Routing",
         "Data":{
            "UnicastRoutingTable":{

            },
            "DistributionPolicy":{

            },
            "EndpointInstances":{

            },
            "Publishers":{

            },
            "EnforceBestPractices":true
         }
      },
      {
         "Name":"HostInformation",
         "Data":{
            "HostId":"e3342552-4a13-b662-aa77-4461b333c85b",
            "DisplayName":"ANDREAS2015",
            "Properties":{
               "Machine":"ANDREAS2015",
               "ProcessID":"26440",
               "UserName":"andreas.ohlund",
               "PathToExecutable":"C:\\Users\\andreas.ohlund\\AppData\\Local\\JetBrains\\Installations\\ReSharperPlatformVs15_c3a37fcb_000\\JetBrains.ReSharper.TaskRunner.CLR45.x64.exe"
            }
         }
      }
   ]
}
```

are we happy with that? (`Name` and `Data` can be renamed)

Comment (SimonCropp) [reply src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs]: call todictionary on the entries. with name as the key and data as the value. then serialize that dictionary

File: src/NServiceBus.AcceptanceTests/ConfigureEndpointLearningTransport.cs (hunk 2 above)
Comment (andreasohlund) [reply src/NServiceBus.AcceptanceTests/ConfigureEndpointLearningTransport.cs]: Agreed, fixed

**Comment by andreasohlund** [comment]
@SimonCropp addressed your comments

@Particular/nservicebus-maintainers went through this with @timbussmann and while there are diagnostics data missing we figured that this would be good to get in since its the thing blocking RC. We can always add more data during the RC phase and in upcoming minors.

Will go through and define a minimum set of required info before RTM and submit that as a seprate PR. 

In short: This one should be good to to

File: src/NServiceBus.AcceptanceTests/Core/Hosting/When_endpoint_starts.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                .Done(c => c.EndpointsStarted)
+                .Run();
+
+            var endpointName = Conventions.EndpointNamingConvention(typeof(MyEndpoint));
+            var startupDiagnoticsFileName = $"{endpointName}-configuration.json";
+
+            var pathToFile = Path.Combine(basePath, startupDiagnoticsFileName);
+            Assert.True(File.Exists(pathToFile));
+
+            Console.Out.WriteLine(File.ReadAllText(pathToFile));
```
Comment (bording) [inline src/NServiceBus.AcceptanceTests/Core/Hosting/When_endpoint_starts.cs]: Why is this here? Do we want to see this in the test output? If so, this should be `TestContext.WriteLine` instead.

File: src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj [hunk 6]
```diff
@@ -40,6 +40,7 @@
     <_PackageFiles Remove="**\obj\**\*.cs" />
     <_PackageFiles Remove="Core\**\*.cs" />
     <_PackageFiles Remove="AssemblyInfo.cs" />
+    <_PackageFiles Remove="EndpointTemplates\DefaultServerWithDiagnostics.cs" />
```
Comment (bording) [inline src/NServiceBus.AcceptanceTests/NServiceBus.AcceptanceTests.csproj]: Why this been removed from the package? Would a downstream want to be able to test an addition to the diagnostics?

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/EndpointDiagnosticSettingsExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+﻿namespace NServiceBus
+{
+    using Settings;
+
+    /// <summary>
+    /// Provides an API to add startup diagnostics.
+    /// </summary>
+    static class EndpointDiagnosticSettingsExtensions
```
Comment (bording) [inline src/NServiceBus.Core/Hosting/StartupDiagnostics/EndpointDiagnosticSettingsExtensions.cs]: Is this supposed to be public so downstreams can use it? If not, why have the xml comments?

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+            if (!Directory.Exists(diagnosticsRootPath))
+            {
+                Directory.CreateDirectory(diagnosticsRootPath);
+            }
+
+            var endpointName = settings.EndpointName();
+
+            // Once we have the proper hosting model in place we can skip the endpoint name since the host would
+            // know how to handle multi hosting but for now we do this so that multi-hosting users will get a file per endpoint
+            var startupDiagnosticsFileName = $"{endpointName}-configuration.json";
```
Comment (bording) [inline src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs]: This naming scheme means that we'll only ever have a single file. Should we consider some sort of multiple file/rolling scheme so that we can collect data for the last N endpoint starts?

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs [hunk 9]
```diff
@@ -45,6 +48,12 @@ public DefaultServer(List<Type> typesToInclude)
 
             configurationBuilderCustomization(configuration);
 
+            if (!EnableDiagnostics)
```
Comment (bording) [inline src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs]: If the new field was `DisableDiagnostics` instead, then this logic would a bit easier to follow, since it would be  "if disabled, then do action to disable".

File: src/NServiceBus.Core.Tests/Logging/DefaultFactoryTests.cs [hunk 10]
```diff
@@ -22,7 +22,7 @@ public void When_directory_is_bad_should_throw()
         [Test]
         public void When_not_running_ASP_NET_should_choose_BaseDirectory_as_logging_directory()
         {
-            var directory = DefaultFactory.FindDefaultLoggingDirectory();
+            var directory = Host.GetOutputDirectory();
```
Comment (bording) [inline src/NServiceBus.Core.Tests/Logging/DefaultFactoryTests.cs]: Does this test really belong in `DefaultFactoryTests` any more now that the logic has moved?

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs (hunk 8 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs]: That was considered but scoped out for now since its really only the latest file we're interested in. Writing to disk is the breaking change, we can always improve on this by doing rolling files, hashes etc in a minor

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/EndpointDiagnosticSettingsExtensions.cs (hunk 7 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Hosting/StartupDiagnostics/EndpointDiagnosticSettingsExtensions.cs]: Good catch, fixed

**Comment by andreasohlund** [comment]
@bording addressed your comments

@timbussmann ready for your review

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        protected internal override void Setup(FeatureConfigurationContext context)
+        {
+            var settings = context.Settings;
+
+            var diagnosticsWriter = GetDiagnosticsWriter(settings);
+
+            context.RegisterStartupTask(new WriteStartupDiagnostics(diagnosticsWriter, settings));
+        }
+
+
```
Comment (timbussmann) [inline src/NServiceBus.Core/Hosting/StartupDiagnostics/HostStartupDiagnostics.cs]: 🔥

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/HostDiagnosticsConfigurationExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    /// Provides diagnostics configuration options.
+    /// </summary>
+    public static class HostDiagnosticsConfigurationExtensions
+    {
+        /// <summary>
+        /// Configures a custom path where host diagnostics is written.
+        /// </summary>
+        /// <param name="config">Configuration object to extend.</param>
+        /// <param name="path">The custom path to use.</param>
+        public static void SetDiagnosticsRootPath(this EndpointConfiguration config, string path)
```
Comment (timbussmann) [inline src/NServiceBus.Core/Hosting/StartupDiagnostics/HostDiagnosticsConfigurationExtensions.cs]: why only the root path? What if I don't want those files be in `.diagnostics`? Not sure this is really an issue, but I'd say usually I expect to be able to customize the whole path if I specify my own?

File: src/NServiceBus.Core/EndpointConfiguration.cs [hunk 13]
```diff
@@ -34,6 +34,7 @@ public EndpointConfiguration(string endpointName)
             pipelineCollection = new PipelineConfiguration();
             Settings.Set<PipelineConfiguration>(pipelineCollection);
             Settings.Set<SatelliteDefinitions>(new SatelliteDefinitions());
+            Settings.Set<StartupDiagnosticEntries>(new StartupDiagnosticEntries());
```
Comment (timbussmann) [inline src/NServiceBus.Core/EndpointConfiguration.cs]: if you would use `GetOrCreate` when specifying sections, this line wouldn't be necessary.

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs [hunk 14]
```diff
@@ -45,6 +48,12 @@ public DefaultServer(List<Type> typesToInclude)
 
             configurationBuilderCustomization(configuration);
 
+            if (DisableDiagnostics)
+            {
+                //no-op diagnostics
+                configuration.GetSettings().Set("NServiceBus.HostStartupDiagnostics", FeatureState.Disabled);
```
[the remark this answers is not in the document]
Comment (bording) [reply src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs]: If they don't want the diagnostic to run, it seems like they'd had to if they aren't using the templates provided here.

File: src/NServiceBus.Core/EndpointConfiguration.cs (hunk 13 above)
Comment (bording) [reply src/NServiceBus.Core/EndpointConfiguration.cs]: `AddStartupDiagnosticsSection` is using `ReadOnlySettings` since it needs to be callable from features, so `GetOrCreate` isn't available.

**Comment by bording** [comment]
Once @timbussmann's comments are addressed, I'd be ready to approve and merge this.

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs [hunk 15]
```diff
@@ -19,6 +20,8 @@ public DefaultServer(List<Type> typesToInclude)
             this.typesToInclude = typesToInclude;
         }
 
+        protected bool DisableDiagnostics = true;
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs]: Do we actually over inherit from DefaultServer? I thought the publisher uses the default server but does not inherit from. Feels a bit strange that this assumes inheritance although we probably never actually do that in the wild.

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServerWithDiagnostics.cs [hunk 16]
```diff
@@ -0,0 +1,10 @@
+﻿namespace NServiceBus.AcceptanceTests.EndpointTemplates
+{
+    class DefaultServerWithDiagnostics : DefaultServer
+    {
+        public DefaultServerWithDiagnostics()
+        {
+            DisableDiagnostics = false;
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServerWithDiagnostics.cs]: Seems that previously we used aggregation instead of inheritance https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultPublisher.cs#L11

File: src/NServiceBus.Core/Hosting/StartupDiagnostics/HostDiagnosticsConfigurationExtensions.cs (hunk 12 above)
Comment (andreasohlund) [reply src/NServiceBus.Core/Hosting/StartupDiagnostics/HostDiagnosticsConfigurationExtensions.cs]: Turns out it works as expected but name was misleading, renamed to `SetDiagnosticsPath`.

**Comment by andreasohlund** [comment]
After talking to @timbussmann we decided to disable diagnostic in the actpt test framework. This means that downstreams will never write diagnostics unless calling

```
    public MyEndpoint()
            {
                EndpointSetup<DefaultServer>()
                    .EnableStartupDiagnostics();
            }
```

cc @danielmarbach

File: src/NServiceBus.AcceptanceTesting/EndpointConfigurationBuilder.cs [hunk 17]
```diff
@@ -59,13 +68,18 @@ public EndpointConfigurationBuilder EndpointSetup<T>(Action<EndpointConfiguratio
...
+                }
+
                 return endpointConfiguration;
             };
 
             return this;
         }
 
-        public EndpointConfigurationBuilder EndpointSetup<T,TContext>(Action<EndpointConfiguration, TContext> configurationBuilderCustomization, Action<PublisherMetadata> publisherMetadata = null) 
+        public EndpointConfigurationBuilder EndpointSetup<T,TContext>(Action<EndpointConfiguration, TContext> configurationBuilderCustomization, Action<PublisherMetadata> publisherMetadata = null)
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTesting/EndpointConfigurationBuilder.cs]: I think you need the logic from the `EndpointSetup` in this overload as well.

File: src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs (hunk 15 above)
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/EndpointTemplates/DefaultServer.cs]: can be removed

**Comment by bording** [comment]
@andreasohlund Looking over the changes, 👍much better end result for the testing setup.
